File: frontend/js/views/liquidity_ratios.js
// Business liquidity ratios, via /calc/liquidity-ratios.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const x = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '×');
export async function renderLiquidityRatios(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.liqr.h1.title">// LIQUIDITY RATIOS</span></h1>
        <p class="muted small" data-i18n="view.liqr.hint.intro">The short-term solvency measures from a company balance sheet: net working capital (current assets − current liabilities), the current ratio, the quick / acid-test ratio (excludes inventory), and the cash ratio. Business-entity ratios, not personal finance. Not advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.liqr.h2.inputs">Balance sheet</h2>
        <form id="liqr-form" class="inline-form">
            <label><span data-i18n="view.liqr.label.company">Company</span><input type="text" name="company_label" value="Acme"></label>
            <label><span data-i18n="view.liqr.label.ca">Current assets ($)</span><input type="number" step="1000" min="0" name="current_assets_usd" value="200000" required></label>
            <label><span data-i18n="view.liqr.label.inv">Inventory ($)</span><input type="number" step="1000" min="0" name="inventory_usd" value="80000"></label>
            <label><span data-i18n="view.liqr.label.cash">Cash & equivalents ($)</span><input type="number" step="1000" min="0" name="cash_and_equivalents_usd" value="50000"></label>
            <label><span data-i18n="view.liqr.label.cl">Current liabilities ($)</span><input type="number" step="1000" min="0" name="current_liabilities_usd" value="100000" required></label>
        </form></div><div id="liqr-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#liqr-form'); const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const gen = async () => {
        const body = { company_label: (form.querySelector('[name="company_label"]').value || '').trim(), current_assets_usd: n('current_assets_usd'), inventory_usd: n('inventory_usd'), cash_and_equivalents_usd: n('cash_and_equivalents_usd'), current_liabilities_usd: n('current_liabilities_usd') };
        try { const d = await api.calcLiquidityRatios(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.liqr.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#liqr-result');
    if (!d.valid) { el.innerHTML = `<p class="muted" data-i18n="view.liqr.invalid">Current liabilities must be positive.</p>`; applyUiI18n(el); return; }
    const statusKey = d.solvent_short_term ? 'view.liqr.solvent' : 'view.liqr.tight';
    const cls = d.solvent_short_term ? 'pos' : 'neg';
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card ${cls}"><div class="label" data-i18n="view.liqr.card.nwc">Net working capital</div><div class="value">${money(d.net_working_capital_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.liqr.card.current">Current ratio</div><div class="value">${x(d.current_ratio)}</div></div>
        <div class="card"><div class="label" data-i18n="view.liqr.card.quick">Quick / acid-test</div><div class="value">${x(d.quick_ratio)}</div></div>
        <div class="card"><div class="label" data-i18n="view.liqr.card.cash">Cash ratio</div><div class="value">${x(d.cash_ratio)}</div></div>
        <div class="card ${cls}"><div class="label" data-i18n="view.liqr.card.status">Short-term</div><div class="value" data-i18n="${statusKey}">${d.solvent_short_term ? 'Solvent' : 'Tight'}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/liquidity_ratios.rs
//! Business liquidity ratios — the short-term solvency measures read off a
//! company balance sheet: net working capital (current assets − current
//! liabilities), the current ratio (current assets ÷ current liabilities), the
//! quick / acid-test ratio (current assets less inventory ÷ current liabilities),
//! and the cash ratio (cash & equivalents ÷ current liabilities). Distinct from
//! the personal-finance ratios module (DTI, emergency fund) — these are
//! business-entity ratios. Pure compute, not financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LiquidityInput {
    pub company_label: String,
    pub current_assets_usd: f64,
    #[serde(default)]
    pub inventory_usd: f64,
    #[serde(default)]
    pub cash_and_equivalents_usd: f64,
    pub current_liabilities_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct LiquidityReport {
    /// Current assets − current liabilities.
    pub net_working_capital_usd: f64,
    /// Current assets ÷ current liabilities.
    pub current_ratio: f64,
    /// (Current assets − inventory) ÷ current liabilities.
    pub quick_ratio: f64,
    /// Cash & equivalents ÷ current liabilities.
    pub cash_ratio: f64,
    /// True when the current ratio is at least 1 (can cover near-term obligations).
    pub solvent_short_term: bool,
    pub valid: bool,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

pub fn generate(i: &LiquidityInput) -> LiquidityReport {
    if i.current_liabilities_usd <= 0.0 {
        return LiquidityReport::default();
    }
    let cl = i.current_liabilities_usd;
    let current = i.current_assets_usd / cl;
    LiquidityReport {
        net_working_capital_usd: cents(i.current_assets_usd - cl),
        current_ratio: round4(current),
        quick_ratio: round4((i.current_assets_usd - i.inventory_usd) / cl),
        cash_ratio: round4(i.cash_and_equivalents_usd / cl),
        solvent_short_term: current >= 1.0,
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> LiquidityInput {
        LiquidityInput {
            company_label: "Acme".into(),
            current_assets_usd: 200_000.0,
            inventory_usd: 80_000.0,
            cash_and_equivalents_usd: 50_000.0,
            current_liabilities_usd: 100_000.0,
        }
    }

    #[test]
    fn standard_ratios() {
        let d = generate(&base());
        assert!(close(d.net_working_capital_usd, 100_000.0));
        assert!(close(d.current_ratio, 2.0));
        assert!(close(d.quick_ratio, 1.2));
        assert!(close(d.cash_ratio, 0.5));
        assert!(d.solvent_short_term);
    }

    #[test]
    fn quick_excludes_inventory() {
        let d = generate(&base());
        let no_inv = generate(&LiquidityInput { inventory_usd: 0.0, ..base() });
        assert!(d.quick_ratio < no_inv.quick_ratio);
        // With no inventory, quick equals current.
        assert!(close(no_inv.quick_ratio, no_inv.current_ratio));
    }

    #[test]
    fn insolvent_when_current_below_one() {
        let d = generate(&LiquidityInput { current_assets_usd: 80_000.0, ..base() });
        assert!(close(d.current_ratio, 0.8));
        assert!(close(d.net_working_capital_usd, -20_000.0));
        assert!(!d.solvent_short_term);
    }

    #[test]
    fn ordering_cash_le_quick_le_current() {
        let d = generate(&base());
        assert!(d.cash_ratio <= d.quick_ratio);
        assert!(d.quick_ratio <= d.current_ratio);
    }

    #[test]
    fn zero_liabilities_invalid() {
        let d = generate(&LiquidityInput { current_liabilities_usd: 0.0, ..base() });
        assert!(!d.valid);
    }
}
